### ISSUE TITLE
`vg deconstruct` (version 1.50.0) crashes with a little PGGB graph

### ISSUE THREAD

**Issue by AndreaGuarracino** [vgteam/vg#4048]
Hi @glennhickey, thank you for taking care of `vg deconstruct` in https://github.com/vgteam/vg/pull/4037, but there are still issues with PGGB graphs. I attach a small graph with which there are no problems with `vg 1.40.0` (the last PGGB-friendly `vg deconstruct` version): [chr6.C4.gfa.gz](https://github.com/vgteam/vg/files/12280581/chr6.C4.gfa.gz).

```shell
gunzip chr6.C4.gfa.gz 
vg deconstruct chr6.C4.gfa

terminate called after throwing an instance of 'std::runtime_error'
  what():  Reference path must have a sample name
━━━━━━━━━━━━━━━━━━━━
Crash report for vg v1.50.0 "Monopoli"
Stack trace (most recent call last):
#24   Object "/home/guarracino/Desktop/vg", at 0x615dd4, in _start
#23   Object "/home/guarracino/Desktop/vg", at 0x205cbf6, in __libc_start_main
#22   Object "/home/guarracino/Desktop/vg", at 0x205b399, in __libc_start_call_main
#21   Object "/home/guarracino/Desktop/vg", at 0xdf4e7b, in vg::subcommand::Subcommand::operator()(int, char**) const
#20   Object "/home/guarracino/Desktop/vg", at 0xcb723b, in main_deconstruct(int, char**)
#19   Object "/home/guarracino/Desktop/vg", at 0xc7a4a6, in std::_Function_handler<void (std::istream&), vg::io::VPKG::try_load_first<vg::GBZGraph, handlegraph::PathHandleGraph>(std::istream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)::{lambda(std::istream&)#1}>::_M_invoke(std::_Any_data const&, std::istream&)
#18   Object "/home/guarracino/Desktop/vg", at 0xc2bf1a, in vg::io::VPKG::with_putback(std::istream&, std::function<void (std::istream&)> const&)
#17   Object "/home/guarracino/Desktop/vg", at 0xc77ab1, in std::_Function_handler<void (std::istream&), vg::io::VPKG::try_load_first_bare<vg::GBZGraph, handlegraph::PathHandleGraph>(std::istream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool&)::{lambda(std::istream&)#1}>::_M_invoke(std::_Any_data const&, std::istream&)
#16   Object "/home/guarracino/Desktop/vg", at 0xc2bf1a, in vg::io::VPKG::with_putback(std::istream&, std::function<void (std::istream&)> const&)
#15   Object "/home/guarracino/Desktop/vg", at 0xc58318, in std::_Function_handler<void (std::istream&), vg::io::VPKG::try_load_bare<handlegraph::PathHandleGraph>(std::istream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)::{lambda(std::istream&)#1}>::_M_invoke(std::_Any_data const&, std::istream&)
#14   Object "/home/guarracino/Desktop/vg", at 0x1555b4a, in std::_Function_handler<void* (std::istream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), vg::io::register_loader_saver_gfa()::{lambda(std::istream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}>::_M_invoke(std::_Any_data const&, std::istream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
#13   Object "/home/guarracino/Desktop/vg", at 0x14ec2c7, in vg::algorithms::gfa_to_path_handle_graph(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, handlegraph::MutablePathMutableHandleGraph*, vg::algorithms::GFAIDMapInfo*, long)
#12   Object "/home/guarracino/Desktop/vg", at 0x1427d57, in vg::get_input_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::function<void (std::istream&)>)
#11   Object "/home/guarracino/Desktop/vg", at 0x14f5c98, in std::_Function_handler<void (std::istream&), vg::algorithms::gfa_to_path_handle_graph(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, handlegraph::MutablePathMutableHandleGraph*, vg::algorithms::GFAIDMapInfo*, long)::{lambda(std::istream&)#1}>::_M_invoke(std::_Any_data const&, std::istream&)
#10   Object "/home/guarracino/Desktop/vg", at 0x14f44c0, in vg::algorithms::GFAParser::parse(std::istream&)
#9    Object "/home/guarracino/Desktop/vg", at 0x14f2bb4, in vg::algorithms::GFAParser::parse(std::istream&)::{lambda()#4}::operator()() const
#8    Object "/home/guarracino/Desktop/vg", at 0x14f0b74, in std::_Function_handler<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::pair<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::pair<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&), vg::algorithms::add_path_listeners(vg::algorithms::GFAParser&, handlegraph::MutablePathMutableHandleGraph*)::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::pair<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::pair<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)#2}>::_M_invoke(std::_Any_data const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::pair<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::pair<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)
#7    Object "/home/guarracino/Desktop/vg", at 0x5bcbe5, in handlegraph::PathMetadata::create_path_name(handlegraph::PathSense const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long const&, unsigned long const&, std::pair<unsigned long, unsigned long> const&) [clone .cold]
#6    Object "/home/guarracino/Desktop/vg", at 0x1f95158, in __cxa_throw
#5    Object "/home/guarracino/Desktop/vg", at 0x1f94ff6, in std::terminate()
#4    Object "/home/guarracino/Desktop/vg", at 0x1f94f8b, in __cxxabiv1::__terminate(void (*)())
#3    Object "/home/guarracino/Desktop/vg", at 0x5e24a3, in __gnu_cxx::__verbose_terminate_handler() [clone .cold]
#2    Object "/home/guarracino/Desktop/vg", at 0x5e4beb, in abort
#1    Object "/home/guarracino/Desktop/vg", at 0x2073dd5, in raise
#0    Object "/home/guarracino/Desktop/vg", at 0x20a07dc, in __pthread_kill
ERROR: Signal 6 occurred. VG has crashed. Visit https://github.com/vgteam/vg/issues/new/choose to report a bug.
Please include this entire error log in your bug report!
━━━━━━━━━━━━━━━━━━━━
```

**Comment by glennhickey** [comment]
Yeah, this is a regression in vg 1.5.0 that's breaking [minigraph-cactus graphs too](https://github.com/vgteam/vg/issues/4043).  @adamnovak is on the case. (Adam are you planning on making a patch release?)

**Comment by adamnovak** [comment]
Yeah, we have to do a patch release after https://github.com/vgteam/vg/pull/4045 is in. But it failed tests because of some kind of assertion failure in read simulation that I can't explain. So I'm not sure when I'll actually be able to do the release.

**Comment by adamnovak** [comment]
OK, I have a 1.50.1 patch release, which fixes the crash on loading this graph.

It does *not* actually fix this issue though. It just restores the 1.49.0 behavior of crashing *after* loading the graph, with an assertion failure of `i < filled` at `include/bdsg/internal/packed_structs.hpp:948`. So there's still an issue to debug here.

If I try converting the graph to various formats before deconstructing, they all fail with some variation of a thing not being found when trying to access it in the graph. So there must be something wrong with the deconstruct algorithm itself.

**Comment by glennhickey** [comment]
The good news is that this regression seems confined to running `deconstruct` exhaustively (ie not using paths for alleles) which in practice is not very useful. 

In order to deconstruct non-trivial graphs, you need to either give it a GBZ, a GBWT (`-g`) or tell it to use embedded paths `-e`. 

As far as I know, PGGB uses `-e`.  In this case, the patch release appears to work fine

```
vg deconstruct chr6.C4.gfa  -e -H '#' -P chm13
```

will give the expected output on vg 1.40.0 as well as the 1.50.1 (noting that the `-H` option isn't needed in the new version and will trigger a warning). 

So I think the current patch should fix PGGB.  I will fix whatever happened to the exhaustive deconstruct as soon as I can, but even when that's working I don't htink there's a use case for it on PGGB graphs.  So unless I hear otherwise, I don't think it would warrant another patch...

**Comment by AndreaGuarracino** [comment]
Fair and practical enough, thank you for hacking on this!

**Comment by luobosi** [comment]
Respectful developers,

Hello! I have encountered a problem similar to the one mentioned by Adam in the following scenario:

I intend to use pggb to obtain a pan-reference genome graph consisting of 31 strains of *Bacillus subtilis*. This graph will serve as the input for vg to perform mapping on new sequencing samples. I have specified **the output format as a BAM file to use other existing variant calling software.**

I did not encounter any obvious warnings or issues during the pggb construction and vg index creation processes. However, when running giraffe, I encountered the following error:
```
warning[vg::Surjector]: Refusing to perform very large alignment against 8196556 bp strand split subgraph for read SRR11461738.55; suppressing further warnings.
Using fragment length estimate: 172.768 +/- 80.2228
warning[vg::giraffe]: Refusing to perform too-large rescue alignment of 148 bp against 80204 bp dagified subgraph for read SRR11461738.20368 which would use more than 1572864 cells and might exhaust Dozeu's allocator; suppressing further warnings.

vg: src/surjector.cpp:2652: vg::Alignment vg::Surjector::realigning_surject(const PathPositionHandleGraph*, const vg::Alignment&, const path_handle_t&, bool, const std::vector<std::pair<std::pair<__gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, vg::Path> >&, std::pair<handlegraph::step_handle_t, handlegraph::step_handle_t>&, bool, bool, bool) const: Assertion `ref_path_interval.first <= ref_path_interval.second' failed.
ERROR: Signal 6 occurred. VG has crashed. Visit https://github.com/vgteam/vg/issues/new/choose to report a bug.
Stack trace path: /tmp/vg_crash_3WgMD4/stacktrace.txt

$ cat stacktrace.txt
Crash report for vg v1.40.0 "Suardi"
Stack trace (most recent call last) in thread 21356:
#17   Object "", at 0xffffffffffffffff, in
#16   Object "miniconda3/envs/pggb/bin/vg", at 0x1e16122, in __clone
#15   Object "miniconda3/envs/pggb/bin/vg", at 0x138d508, in start_thread
#14   Object "miniconda3/envs/pggb/bin/vg", at 0x1d4aadd, in gomp_thread_start
#13   Object "miniconda3/envs/pggb/bin/vg", at 0xcca12f, in unsigned long vg::io::paired_for_each_parallel_after_wait<vg::Alignment>(std::function<bool (vg::Alignment&, vg::Alignment&)>, std::function<void (vg::Alignment&, vg::Alignment&)>, std::function<bool ()>, unsigned long) [clone ._omp_fn.0]
#12   Object "miniconda3/envs/pggb/bin/vg", at 0x1d4d3b7, in gomp_team_barrier_wait_end
#11   Object "miniconda3/envs/pggb/bin/vg", at 0x1d44a7b, in gomp_barrier_handle_tasks
#10   Object "miniconda3/envs/pggb/bin/vg", at 0xcca2da, in unsigned long vg::io::paired_for_each_parallel_after_wait<vg::Alignment>(std::function<bool (vg::Alignment&, vg::Alignment&)>, std::function<void (vg::Alignment&, vg::Alignment&)>, std::function<bool ()>, unsigned long) [clone ._omp_fn.1]
#9    Object "miniconda3/envs/pggb/bin/vg", at 0xbedb24, in std::_Function_handler<void (vg::Alignment&, vg::Alignment&), main_giraffe(int, char**)::{lambda()#1}::operator()() const::{lambda(vg::Alignment&, vg::Alignment&)#6}>::_M_invoke(std::_Any_data const&, vg::Alignment&, vg::Alignment&)
#8    Object "miniconda3/envs/pggb/bin/vg", at 0x12fa64b, in vg::io::AlignmentEmitter::emit_mapped_pair(std::vector<vg::Alignment, std::allocator<vg::Alignment> >&&, std::vector<vg::Alignment, std::allocator<vg::Alignment> >&&, long)
#7    Object "miniconda3/envs/pggb/bin/vg", at 0x10fbcdf, in vg::SurjectingAlignmentEmitter::emit_mapped_pairs(std::vector<std::vector<vg::Alignment, std::allocator<vg::Alignment> >, std::allocator<std::vector<vg::Alignment, std::allocator<vg::Alignment> > > >&&, std::vector<std::vector<vg::Alignment, std::allocator<vg::Alignment> >, std::allocator<std::vector<vg::Alignment, std::allocator<vg::Alignment> > > >&&, std::vector<long, std::allocator<long> >&&)
#6    Object "miniconda3/envs/pggb/bin/vg", at 0x111dbf5, in vg::Surjector::surject(vg::Alignment const&, std::unordered_set<handlegraph::path_handle_t, std::hash<handlegraph::path_handle_t>, std::equal_to<handlegraph::path_handle_t>, std::allocator<handlegraph::path_handle_t> > const&, bool, bool) const
#5    Object "miniconda3/envs/pggb/bin/vg", at 0x111c562, in vg::Surjector::surject_internal(vg::Alignment const*, vg::multipath_alignment_t const*, vg::Alignment*, vg::multipath_alignment_t*, std::unordered_set<handlegraph::path_handle_t, std::hash<handlegraph::path_handle_t>, std::equal_to<handlegraph::path_handle_t>, std::allocator<handlegraph::path_handle_t> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, long&, bool&, bool, bool) const
#4    Object "miniconda3/envs/pggb/bin/vg", at 0x11113bd, in vg::Surjector::realigning_surject(handlegraph::PathPositionHandleGraph const*, vg::Alignment const&, handlegraph::path_handle_t const&, bool, std::vector<std::pair<std::pair<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, vg::Path>, std::allocator<std::pair<std::pair<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, vg::Path> > > const&, std::pair<handlegraph::step_handle_t, handlegraph::step_handle_t>&, bool, bool, bool) const
#3    Object "miniconda3/envs/pggb/bin/vg", at 0x1d6e035, in __assert_fail
#2    Object "miniconda3/envs/pggb/bin/vg", at 0x58c607, in __assert_fail_base.cold
#1    Object "miniconda3/envs/pggb/bin/vg", at 0x58c737, in abort
#0    Object "miniconda3/envs/pggb/bin/vg", at 0x13908eb, in raise
```
I tried **changing the fastq data, but the same error persists**. I checked the fastq content based on the provided warning location and did not find any abnormalities. Therefore, I suspect that this issue may be related to the gfa file constructed by pggb or it might be a problem with the vg software itself.

While investigating this issue, I found that this seems to be a recent error and it was also mentioned that the minigraph-cactus graph has similar problems. I noticed that you attempted to resolve it using the **vg deconstruct** command. However I have not utilized the deconstruct command in my commands.

I would like to know how to resolve this error and **obtain the correct BAM file**. Additionally, I am curious if this approach of obtaining a BAM file and subsequently using traditional variant calling software such as GATK is **theoretically feasible** ? Lastly, is there any difference in content between directly obtaining a BAM file using the **-o BAM** provided command and obtaining a GAM file first, then converting it to BAM using **vg view**?

Below are the commands and software versions I used:
```
# pggb version
pggb 0.5.4

# vg version
vg: variation graph tool, version v1.40.0 "Suardi"

# Command
pggb -i pggb/test/test.fna.gz -o pggb/test/out/ -n 31 -t 32 -p 90 -s 5000

vg autoindex --workflow giraffe -t 16 -p pggb/test/out/bs31 \
 -g pggb/test/out/test.fna.gz.225d33b.417fcdf.a1fd1f6.smooth.final.gfa

vg giraffe -t 32 -Z pggb/test/out/bs31.giraffe.gbz \
 -m pggb/test/out/bs31.min -d pggb/test/out/bs31.dist \
 -f pggb/test/out/SRR11461738_1.fastq \
 -f pggb/test/out/SRR11461738_2.fastq \
 -p -o BAM >pggb/test/out/bs31.bam
```
I **apologize** for my lack of knowledge and asking multiple questions. I am eagerly looking forward to your help in resolving my issues and confusion. If you need data to reproduce the problem, please let me know your email address.Thank you very much!

**Comment by glennhickey** [comment]
@luobosi Try the [latest vg release](https://github.com/vgteam/vg/releases/tag/v1.50.1) for both your vg index construction and mapping. 

Note that your crash has nothing to do with this github issue about `vg deconstruct`, but many PGGB / surjection issues have been fixed between v1.40.0 and now.
